[ART-1728] Create `build/olm_bundle` job

thiagoalessio opened the pull request:
Related story: [ART-1728](https://issues.redhat.com/browse/ART-1728)

### What is introduced by this PR

A new job, called `build/olm_bundle` that builds bundle containers for OLM Operators, either attached to an advisory or the latest build found.

Test runs are happening at https://saml.buildvm.openshift.eng.bos.redhat.com:8888/job/hack/job/talessio-aos-cd-jobs/job/ART-1728

sosiouxme commented on jobs/build/olm_bundle/olm_bundles.groovy:
```diff
@@ -0,0 +1,69 @@
+def doozer(cmd) {
+    commonlib.shell(
+        script:"""
+            python3 /home/jenkins/talessio/doozer/doozer \
```
obviously i won't merge it like this :-D

doozer PR is merged but not yet deployed on buildvm.

sosiouxme commented on jobs/build/olm_bundle/Jenkinsfile:
```diff
@@ -0,0 +1,125 @@
+            defaultValue: ''
+        )
+        string(
+            name: 'BUNDLES_ADVISORY',
```
I don't think we'll want a separate advisory for bundles. We'll just put them in the same advisory the rest of extras go in.

I can't think of a reason we'd want to run this job and not attach the results anywhere (other than... testing the job?).

Maybe the default could be to attach them to current extras, and this parameter could be to explicitly send them somewhere else (e.g. rebuilding something for an in-flight-but-not-latest release that wasn't right initially).

sosiouxme commented on jobs/build/olm_bundle/Jenkinsfile:
```diff
@@ -0,0 +1,125 @@
+                        bundle_packages = operator_packages.collect {
+                            it.replace('-operator', '-operator-bundle')
+                        }
+                        bundle_nvrs = olm_bundles.get_latest_builds(bundle_packages)
```
i'm think we want per-version locking around this. otherwise it's conceivable that distgit updates could collide.

(but don't use the same lock as is being used for the non-bundle image updates)

sosiouxme commented on jobs/build/olm_bundle/Jenkinsfile:
```diff
@@ -0,0 +1,125 @@
+            }
+            steps {
+                script {
+                    olm_bundles.attach_to_bundles_advisory(bundle_nvrs)
```
i think we want a lock around this too since it changes advisory state

thiagoalessio commented:
Thanks a lot for the feedback @sosiouxme ! I tried to address all your comments. Just not 100% sure if I correctly understood you meant [here](https://github.com/openshift/aos-cd-jobs/pull/2232#discussion_r433049116).

Now my implementation is supposed to do the following:
- If `EXTRAS_ADVISORY` is informed, we fetch OLM Operator NVRs from there _and_ attach corresponding bundles back at the same advisory;
- If `EXTRAS_ADVISORY` is empty, we fetch OLM Operator NVRs from `brew latest-build` and just build bundles without attaching them anywhere.

I should have created separated commits to facilitate review, and just squash them on merge, but I just `git commit --amend` :( ... so here is a diff before/after addressing your comments:

```
 $ git diff 3f5d3746234e47100b658ba2a1148c9ae8a3e649 3d5ea96469b0ed721abc941271e575d507d1eedd -- jobs/build/olm_bundle/
```
```diff
diff --git a/jobs/build/olm_bundle/Jenkinsfile b/jobs/build/olm_bundle/Jenkinsfile
index 5a5d54b31..3921dc5b3 100644
--- a/jobs/build/olm_bundle/Jenkinsfile
+++ b/jobs/build/olm_bundle/Jenkinsfile
@@ -39,15 +39,10 @@ pipeline {
                gets solved */
         )
         string(
-            name: 'OLM_OPERATOR_ADVISORY',
-            description: '(Optional) Fetch OLM Operators NVRs from advisory.\n' +
-                         'Leave empty to fetch NVRs from "brew latest-build"',
-            defaultValue: ''
-        )
-        string(
-            name: 'BUNDLES_ADVISORY',
-            description: '(Optional) Attach built bundles to advisory.\n' +
-                         'Leave empty to just build without attaching them anywhere',
+            name: 'EXTRAS_ADVISORY',
+            description: '(Optional) Fetch OLM Operators NVRs from advisory.\n'  +
+                         'Leave empty to fetch NVRs from "brew latest-build".\n' +
+                         'Bundles won\'t be attached to any advisory if empty.',
             defaultValue: ''
         )
         booleanParam(
@@ -62,8 +57,7 @@ pipeline {
             steps {
                 script {
                     currentBuild.displayName += " (${params.BUILD_VERSION})"
-                    currentBuild.description  = "${params.OLM_OPERATOR_ADVISORY ?: '-'} "
-                    currentBuild.description += "${params.BUNDLES_ADVISORY ?: '-'}"
+                    currentBuild.description  = "${params.EXTRAS_ADVISORY ?: '-'} "
 
                     def only = olm_bundles.commonlib.cleanCommaList(params.ONLY).split(',')
                     def exclude = olm_bundles.commonlib.cleanCommaList(params.EXCLUDE).split(',')
@@ -73,7 +67,7 @@ pipeline {
         }
         stage('Get advisory builds') {
             when {
-                expression { ! params.OLM_OPERATOR_ADVISORY.isEmpty() }
+                expression { ! params.EXTRAS_ADVISORY.isEmpty() }
             }
             steps {
                 script {
@@ -85,7 +79,7 @@ pipeline {
         }
         stage('Get latest builds') {
             when {
-                expression { params.OLM_OPERATOR_ADVISORY.isEmpty() }
+                expression { params.EXTRAS_ADVISORY.isEmpty() }
             }
             steps {
                 script {
@@ -99,25 +93,29 @@ pipeline {
             }
             steps {
                 script {
-                    if (params.DRY_RUN) {
-                        bundle_packages = operator_packages.collect {
-                            it.replace('-operator', '-operator-bundle')
+                    lock("olm_bundle-${params.BUILD_VERSION}") {
+                        if (params.DRY_RUN) {
+                            bundle_packages = operator_packages.collect {
+                                it.replace('-operator', '-operator-bundle')
+                            }
+                            bundle_nvrs = olm_bundles.get_latest_builds(bundle_packages)
+                            print(bundle_nvrs.join('\n'))
+                            return
                         }
-                        bundle_nvrs = olm_bundles.get_latest_builds(bundle_packages)
-                        print(bundle_nvrs.join('\n'))
-                        return
+                        bundle_nvrs = olm_bundles.build_bundles(operator_nvrs)
                     }
-                    bundle_nvrs = olm_bundles.build_bundles(operator_nvrs)
                 }
             }
         }
         stage('Attach bundles to advisory') {
             when {
-                expression { ! params.BUNDLES_ADVISORY.isEmpty() }
+                expression { ! params.EXTRAS_ADVISORY.isEmpty() }
             }
             steps {
                 script {
-                    olm_bundles.attach_to_bundles_advisory(bundle_nvrs)
+                    lock("olm_bundle-${params.BUILD_VERSION}") {
+                        olm_bundles.attach_to_bundles_advisory(bundle_nvrs)
+                    }
                 }
             }
         }
diff --git a/jobs/build/olm_bundle/olm_bundles.groovy b/jobs/build/olm_bundle/olm_bundles.groovy
index 4e8992e0c..c199a333e 100644
--- a/jobs/build/olm_bundle/olm_bundles.groovy
+++ b/jobs/build/olm_bundle/olm_bundles.groovy
@@ -1,20 +1,8 @@
 buildlib = load('pipeline-scripts/buildlib.groovy')
 commonlib = buildlib.commonlib
 
-/*
- * Temporary wrapper to use my fork of doozer until
- * https://github.com/openshift/doozer/pull/198 gets merged
- */
 def doozer(cmd) {
-    commonlib.shell(
-        script:"""
-            python3 /home/jenkins/talessio/doozer/doozer \
-            --working-dir ${env.WORKSPACE}/workDir \
-            -g openshift-${params.BUILD_VERSION} \
-            ${cmd}
-        """,
-        returnStdout: true
-    )
+    buildlib.doozer("-g openshift-${params.BUILD_VERSION} ${cmd}", [capture: true])
 }
 
 def elliott(cmd) {
@@ -56,13 +44,13 @@ def build_bundles(operator_nvrs) {
 }
 
 /*
- * Attach given <bundle_nvrs> to <params.BUNDLES_ADVISORY>
+ * Attach given <bundle_nvrs> to <params.EXTRAS_ADVISORY>
  */
-def attach_to_bundles_advisory(bundle_nvrs) {
-    elliott("change-state -s NEW_FILES -a ${params.BUNDLES_ADVISORY} ${params.DRY_RUN ? '--noop' : ''}")
+def attach_bundles_to_extras_advisory(bundle_nvrs) {
+    elliott("change-state -s NEW_FILES -a ${params.EXTRAS_ADVISORY} ${params.DRY_RUN ? '--noop' : ''}")
     flags = bundle_nvrs.collect { "--build ${it}" }.join(' ')
-    elliott("find-builds -k image ${flags} --attach ${params.BUNDLES_ADVISORY}")
-    elliott("change-state -s QE -a ${params.BUNDLES_ADVISORY} ${params.DRY_RUN ? '--noop' : ''}")
+    elliott("find-builds -k image ${flags} --attach ${params.EXTRAS_ADVISORY}")
+    elliott("change-state -s QE -a ${params.EXTRAS_ADVISORY} ${params.DRY_RUN ? '--noop' : ''}")
 
 }
 ```

sosiouxme commented:
@thiagoalessio looks good now. I'd probably allow the user to put "default" in the EXTRAS_ADVISORY field and look it up from ocp-build-data, but that can be a later refinement if release artists see fit :)

sosiouxme commented:
oh, and also... first run should be a "mock run" failing out just to have jenkins define parameters :)
